### PR TITLE
coll/alltoallv,ialltoallv: skip send/recv 0-byte data

### PR DESCRIPTION
**Pull request by wenduwan**
The previous change c1a98f19ea108db1f6e895c5805f47ca318c70ff should have checked for total data size in bytes instead of count. This patch fixes that.

### CODE REVIEW

**Comment by ggouaillardet** [comment]
FWIW, I did skip a lot of zero byte send/recv in the collective several years ago, but the changes were reverted.
The reason is zero byte send/recv provides implicit synchronization, and without it, it is easy to have some processes lag behind and hence have a lot of unexpected messages that will ultimately increase memory overhead and destroy performances.
@bosilca any thoughts on this?

**Comment by wenduwan** [comment]
@ggouaillardet Could you elaborate more on the topic, since I don't know the background of your change earlier.

This change is to fix a bug in IBM mtt `alltoallv_somezeros`. The test exchanges different data on different ranks:
- Some rank sends 0-byte datatype of non-zero count
- Some ranks sned non-zero datatype of 0 count

As a result, before this change the algorithm was posting recvs without matching sends.

The point about synchronization is interesting - curious why not use a barrier in that case?

**Comment by rhc54** [comment]
Hmmm...I'm no MPI expert, but I very much doubt that there is a clause prohibiting one from using alltoall with zero-byte data. Yes, you could do it with a barrier - but that doesn't mean there is a rule against using the other collective that way.

So I suspect this should be fixed a different way.

**Comment by ggouaillardet** [comment]
#3255 is the issue I was talking about.

This is really an implementation choice. When some ranks have nothing to send/receive, one option is to send/receive zero size messages, and an other one is not to send/receive anything.
At first glance, the latter looks like a good idea, but because Open MPI does not do any kind of control flow, that can lead to undesirable behavior caused by unexpected messages. The `coll/sync` module (kind of) addresses this.

Long story short, we've been there before, and it turned out that was not a good idea after all.

**Comment by wenduwan** [comment]
Thanks both. I think this is worthy more discussion.

The alternative is to always send and recv 0-byte messages. Currently I see many places that skip 0-count but not 0-byte data, which will also lead to unexpected messages in the network layer as you pointed out. However, if unexpected message is the only concern, I imagine that would be equivalent to not send/recv 0-byte messages.

Regarding the synchronization behavior, I'm not sure if it is an accidental side effect - it is not obvious to me that someone can rely on 0-byte message to synchronize...

**Comment by wenduwan** [comment]
@bosilca Could you weigh in on this issue?

**Comment by rhc54** [comment]
Perhaps it is just me, but I have encountered quite a few apps that sync'd themselves by doing a collective where everyone passed zero-byte data. Yeah, they could use a barrier - but the expectation was that a zero-byte collective would still act as a collective. The collective was used in a part of the code that might - depending on state of the app - involve non-zero-byte data, but they were guaranteed to sync at that point regardless.

🤷‍♂️  not sure what you gain by not sending them, but it's up to you folks. Just making you aware of what we've seen out there.

**Comment by wenduwan** [comment]
Come to think more about it, I believe we need to choose between:
- Skip send and recv if count is 0 AND datatype size is 0
- Always send and recv if count is 0 OR datatype size is 0

The current logic that only checks for count but not datatype size is inefficient and leads to unexpected messages downstream.

The not-so-clear thing to me is whether the application's expectation is reasonable - does MPI standard provide any guarantee on collective synchronization behavior, especially when the message is 0-byte?

**Comment by devreal** [comment]
> The not-so-clear thing to me is whether the application's expectation is reasonable - does MPI standard provide any guarantee on collective synchronization behavior, especially when the message is 0-byte?

It does not. Chapter 6 of MPI 4.1 states:

> Collective operations can (but are not required to) complete as soon as the caller’s
participation in the collective communication is finished. A blocking operation is complete
as soon as the call returns. [...] The completion of a collective operation indicates that the caller
is free to modify locations in the communication buffer. It does not indicate that other
MPI processes in the group have completed or even started the operation (unless otherwise
implied by the description of the operation). Thus, a collective communication operation
may, or may not, have the effect of synchronizing all participating MPI processes.

The "unless otherwise implied by the description of the operation" part only applies to `MPI_Barrier` (obviously) and some of the RMA operations. Any application *relying* on any other collective communication operation to be synchronizing is erroneous.

**Comment by bosilca** [comment]
As indicated above, the only MPI-blessed synchronization primitive is the barrier. While OMPI put some efforts into helping, technically incorrect applications, MPICH never did, so whatever applications @rhc54 refers to never worked in a non-OMPI setup. These applications should be fixed (instead of expecting an MPI implementation to provide a sketchy workaround).

Moreover, MPI standard is clear that the arguments for collective communications should match at all participants. Which means that unlike point-to-point, where it is legal for a receiver to receive less than expected and thus all matching packets must be sent, for collective communications MPI implementations have the choice of sending or not a synchronization packet. @ggouaillardet is correct, our initial choice was to always send 0-bytes messages, but clearly over time, and with the increase in the number of developers in the collective communication components, this requirement got relaxed. 

Correct MPI applications should only use 0-bytes datatypes for generalized collective operation (`_v` and `_w`) as a replacement for neighborhood collectives (where the topology changes during the runtime). In such cases, it seems unlikely that particular processes get excluded indefinitely from the collective process, and as a result they will naturally synchronize. Thus, we can safely skip all 0-bytes communications. However, if we choose this route we need to audit the collective code and as indicated by @wenduwan we need to ensure we use the size of the messages and not the count or the datatype size.

**Comment by ggouaillardet** [comment]
I think we all agree here the MPI standard does **not** mandate a barrier semantic in `MPI_Alltoallv()` and many other subroutines.

The issue here is that Open MPI does not do any control flow (underlying libraries such as `UCX` or `libfabric` might do so though) so it is pretty easy for a correct MPI application to have a some ranks bomb others with lots of unexpected messages, causing massive memory overhead that will result in performance degradation and/or crash.
This is what was reported in #3255 and the reason why the changes I previously implemented (e.g. never send zero size messages unless we have to) were finally reverted.

**Comment by rhc54** [comment]
FWIW, this is all starting to page its way back into my memory. Been a long time - the problem was originally encountered back at LANL when we moved to RoadRunner. We had at least one proc that was always a little slower than the rest, and lots and lots of apps were failing as the unexpected message queue would explode...which is why I wrote the coll/sync module to introduce a periodic barrier. Not the prettiest solution, but it worked (and still does today).

Also FWIW: as I recall, MPICH couldn't even run on that machine for those apps. I don't know if/what they have done about it, but at least back then it just crashed/burned due to unexpected messages. Short of hard coding barriers, there didn't appear to be any other solution.

**Comment by wenduwan** [comment]
Thank you all for chiming in.

Looking closer at https://github.com/open-mpi/ompi/pull/3285 I notice that `mca_coll_base_alltoallv_intra_basic_inplace` was doing the wrong thing - it only checked for count but not size. I wonder if that was the reason for the unexpected messages, and by removing the 0-count check we forced send/recv for all message sizes. Based on the above discussions, that worked as expected.

This change should do the same thing from the opposite direction by not send/recv'ing 0-size messages. If unexpected messages are the primary concern here, I think this PR should make sense - it avoids sending messages and therefore should reduce unexpected messages overall.

I also agree with @bosilca that we should audit the `*v` collectives in general and ensure each send/recv pair are doing the matching send/recv ops based on the data size instead of count.

**Comment by wenduwan** [comment]
@bosilca @ggouaillardet @devreal I think we need a consensus for this change.

As the algorithm currently stands, it will result in many unexpected messages, which this PR should fix. I also see similar problems in other collectives that can/should be fixed separately.

Do you have more concerns?

**Comment by wenduwan** [comment]
@bosilca Thanks for the catch! I have refactored the change. Specifically, instead of checking datatype size for each send/recv, I figured we can return success early in alltoall.

**Review by bosilca — APPROVED** [review]

File: ompi/mca/coll/libnbc/nbc_ialltoallv.c [hunk 1]
```diff
@@ -48,6 +48,7 @@ static int nbc_alltoallv_init(const void* sendbuf, const int *sendcounts, const
                               mca_coll_base_module_t *module, bool persistent)
 {
   int rank, p, res;
+  size_t sndsz, rcvsz;
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply ompi/mca/coll/libnbc/nbc_ialltoallv.c]: Thank you! Fixed now.
